### ISSUE TITLE
Watch mode repeatedly prints `Error returned from file system watcher.` when virtual env is inside the directory being analyzed

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#5415]
**Describe the bug**
When running `pyright -w` on the command line, it correctly reports some errors then infinitely loops with `Error returned from file system watcher.` I'm on MacOS 13.4.1 (22F82). Happens both in iTerm2 and in the terminal within PyCharm. Perhaps it's at least worth exposing the error [here](https://github.com/microsoft/pyright/blob/c767228a82535b0bdd404480d6ffd5a1d24c8884/packages/pyright-internal/src/common/chokidarFileWatcherProvider.ts#L60)? I did so manually locally and see `Error: EMFILE: too many open files, watch`. It also only happens on larger projects, so [this comment](https://github.com/microsoft/pyright/blob/c767228a82535b0bdd404480d6ffd5a1d24c8884/packages/pyright-internal/src/common/chokidarFileWatcherProvider.ts#L42-L43) might be relevant. 

UPDATE: I verified that if I manually change [this line](https://github.com/microsoft/pyright/blob/c767228a82535b0bdd404480d6ffd5a1d24c8884/packages/pyright-internal/src/common/chokidarFileWatcherProvider.ts#L47) to 
```typescript
const excludes: string[] = ["**/node_modules/**", "**/__pycache__/**", "**/.*/**"];
```

then the problem is resolved. I see that plumbing through `configOptions.exclude` to the file system watcher constructor is not trivial, but hopefully someone more familiar with the code can do it quickly. 

**VS Code extension or command-line**
pyright 1.1.316 from the command line. 

**Additional context**
Only instance of the error message I could find on a Google search was [this bug](https://github.com/microsoft/pylance-release/issues/936).

**Comment by adampauls** [comment]
Is there any chance that the pyright excludes mentioned [here](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#main-configuration-options) are not piped through to the file watcher [here](https://github.com/microsoft/pyright/blob/c767228a82535b0bdd404480d6ffd5a1d24c8884/packages/pyright-internal/src/common/chokidarFileWatcherProvider.ts#L47)? That would make sense, since we have a `.venv` folder inside the folder that we are running `pyright` on.

**Comment by erictraut** [comment]
Piping through the excludes is not a proper solution. Excluded files and directories may still be imported by included files, so they should be retained as part of the files system watcher. The problem here is that MacOS doesn't support recursive file system watchers like Linux and Windows do.

**Comment by erictraut** [comment]
Piping through the excludes is not a proper solution. Excluded files and directories may still be imported by included files, so they should be retained as part of the files system watcher. The problem here is that MacOS doesn't support recursive file system watchers like Linux and Windows do.

**Comment by erictraut** [comment]
Piping through the excludes is not a proper solution. Excluded files and directories may still be imported by included files, so they should be retained as part of the files system watcher. The problem here is that MacOS doesn't support recursive file system watchers like Linux and Windows do.

**Comment by adampauls** [comment]
Upon further investigation, it looks like the problem could be that `canUse()` is returning `undefined` [here](https://github.com/paulmillr/chokidar/blob/08318d62f1fa731122d2b8853400f1b3252db8c9/index.js#L325). I added some further logging to chokidar and see that I'm getting 

```
Error: Cannot find module 'fsevents'
Require stack:
- /Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/pyright.js
- /Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/index.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1075:15)
    at Module._load (node:internal/modules/cjs/loader:920:27)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at 4943 (/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/pyright.js:1:84)
    at i (/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/pyright.js:1:816)
    at 5047 (/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/vendor.js:2:464020)
    at i (/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/pyright.js:1:816)
    at 8590 (/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/vendor.js:2:451691)
    at i (/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/pyright.js:1:816) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/dist/pyright.js',
    '/Users/adpauls/.cache/pyright-python/1.1.316/node_modules/pyright/index.js'
  ]
}
```

I don't have much TS experience so I don't know what's going on here. I do see that `fsevents` is an [optional dependency](https://github.com/paulmillr/chokidar/blob/08318d62f1fa731122d2b8853400f1b3252db8c9/package.json#L26) from Chokidar's point of view. Is it perhaps `pyright`'s responsibility to make sure that library is installed on MacOS?

**Comment by adampauls** [comment]
Ah, this is an issue in [the Python wrapper's](https://github.com/RobertCraigie/pyright-python]) right? I will report this issue over there.

**Comment by erictraut** [comment]
I was incorrect above when I said that "MacOS doesn't support recursive file system watchers". It does support recursive file watching via the "fsevents" API. Chokidar uses this API on MacOS to implement efficient recursive file watching events.

I think you may be correct that chokidar is failing to use "fsevents" on your system and is falling back on a technique that doesn't scale well to large file hierarchies. 

Why do you think this is related to the Python wrapper? Does the problem go away if you install pyright using npm (`npm install pyright`)?

**Comment by adampauls** [comment]
I didn't try installing through npm, but I did say [in the other https://github.com/RobertCraigie/pyright-python/issues/191) that running `npm install fsevents` in the directory that the Python wrapper installs pyright to fixed the issue. I will double check that installing via npm works as well, but I expect it will.

**Comment by erictraut** [comment]
The python wrapper simply calls `npm install pyright`, so I would be surprised if you saw different behaviors between the two. The `npm install pyright` should install `fsevents` automatically. I'll need to investigate further to determine why it might fail to install.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.318, which I just published. Thanks @adampauls for your help in identifying a fix!

**Comment by adampauls** [comment]
Confirmed it is now fixed with me for the released version, thanks!
